exercises/rest-api/canonical-data.json

```diff
@@ -5,12 +5,10 @@
       "The state of the API database before the request is represented",
       "by the input->database object. Your track may determine how this",
       "initial state is set.",
-      "The inputs and outputs are represented as arrays of strings to",
-      "improve readability in this JSON file.",
-      "Your track may choose whether to present the input as a single",
-      "string (concatenating all the lines) or as the list.",
-      "In most cases, it seems to make sense to expect the output as",
-      "a single string.",
+      "The input->payload object and the output are represented as arrays",
+      "of strings to improve readability in this JSON file. However, in ",
+      "track implementations, it seems to make sense to expect them as a ",
+      "single string.",
```

Comment:
@ErikSchierboom thoughts on the wording?

Reply:
Maybe we should completely circumvent this issue by just using JSON to represent the payload. So instead of:

```json
"payload": [
 "{",
  "  \"user\": \"Adam\"",
  "}"
]
```

why not do:

```json
"payload": {
  "user": "Adam"
}
```

Reply:
I suppose if we are moving away from a focus on string parsing that makes sense. It would then be up to tracks to determine how to represent that JSON data.

Reply:
That's a very good idea, but there should be a note to implementers somewhere that the payload should be marshalled to a string before being tested. Parsing and validating the payload is a non-trivial part of the whole problem of building a REST API.

Reply:
It's just my opinion that we should be focusing on the HTTP/REST part, but I do think that makes sense as we already have lots of string parsing exercises but nothing HTTP-related. That's also why I think this is a fantastic new exercise!

Reply:
I'd go farther and suggest that the `expect` objects are also represented in the canonical data as raw JSON, with it being the responsibility of the exercise implementer to parse the serialized JSON string which the student should provide and determine equivalence with the canonical expected data.

Reply:
Porque no los dos?

You're right that there are a bunch of string parsing exercises and nothing focusing on HTTP traffic, but this exercise will probably be one of the more advanced in most tracks. I think it's legitimate in a late exercise to require several skills to operate in parallel.

Reply:
> Parsing *and validating* the payload...

*(Emphasis mine)*

On the subject, I have been considering the idea of adding additional tests that would force the user's implementation to validate URLs and payloads as well. Is there an argument to not add such tests?

Reply:
I don't see any argument against testing validation of URLs and payloads. The argument in favor seems pretty obvious: that's something that real HTTP servers actually have to do reliably.

Reply:
> On the subject, I have been considering the idea of adding additional tests that would force the user's implementation to validate URLs and payloads as well. Is there an argument to not add such tests?

I would think this would be quite important, so feel free to add them,

Reply:
> I'd go farther and suggest that the expect objects are also represented in the canonical data as raw JSON, with it being the responsibility of the exercise implementer to parse the serialized JSON string which the student should provide and determine equivalence with the canonical expected data.

Agreed!